source/common/router/router.cc

```diff
@@ -211,18 +211,19 @@ Http::FilterHeadersStatus Filter::decodeHeaders(Http::HeaderMap& headers, bool e
   }
 
   // Determine if there is a direct response for the request.
-  if (route_->directResponseEntry()) {
-    auto response_code = route_->directResponseEntry()->responseCode();
-    if (response_code >= Http::Code::MultipleChoices && response_code < Http::Code::BadRequest) {
-      config_.stats_.rq_redirect_.inc();
-      Http::Utility::sendRedirect(*callbacks_, route_->directResponseEntry()->newPath(headers),
-                                  response_code);
-      return Http::FilterHeadersStatus::StopIteration;
-    }
+  const auto* direct_response = route_->directResponseEntry();
+  if (direct_response != nullptr) {
     config_.stats_.rq_direct_response_.inc();
-    sendLocalReply(route_->directResponseEntry()->responseCode(),
-                   route_->directResponseEntry()->responseBody(), false);
-    // TODO(brian-pane) support sending response_headers_to_add.
+    sendLocalReply(
+        direct_response->responseCode(), direct_response->responseBody(),
+        [ this, direct_response, &request_headers = headers ](Http::HeaderMap & response_headers)
+            ->void {
+              const auto new_path = direct_response->newPath(request_headers);
+              if (!new_path.empty()) {
+                response_headers.addCopy(Http::LowerCaseString("location"), new_path);
```

Comment:
My previous comment on this still needs action just in case it got lost (switch to `addReferenceKey` and use the constant in the headers class).

Reply:
There isn't a constant for location in the headers class:
```
$ egrep -ir location include/envoy/http
$ echo $?
1
$
```

Reply:
Oh, wait you mean the other headers class, the one in source/common/http.

Reply:
Yes, sorry, was going to provide you link. Basically just look at how it's being used in the other redirect code.